File: Implement strStr().cpp
class Solution {
public:
    vector<int> getNext(char *needle) {
        int i = 0, j = -1;
        vector<int> next(strlen(needle));
        next[0] = -1;
        
        while (i < strlen(needle)) {
            if (j == -1 || needle[i] == needle[j]) {
                i++;
                j++;
                next[i] = j;
            } else {
                j = next[j];
            }
        }
        return next;
    }
    char *strStr(char *haystack, char *needle) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int i = 0, j = 0;
        if (strlen(needle) == 0) return haystack;
        vector<int> next = getNext(needle);
        
        while (i < strlen(haystack) && j < strlen(needle)) {
            if (j == -1 || haystack[i] == needle[j]) {
                i++;
                j++;
            } else {
                j = next[j];
            }
        }
        
        if (j == strlen(needle)) {
            return haystack + i - j;
        } else {
            return NULL;
        }
        
    }
};

File: Implement strStr().cs
public class Solution {
    public int StrStr(string haystack, string needle) {
        if (haystack == null || string.IsNullOrEmpty(needle)) {
            return 0;
        }
        
        if (needle.Length > haystack.Length) {
            return -1;
        }
        
        for (var i = 0; i <= haystack.Length - needle.Length; i++) {
            var j = 0;
            for (j = 0; j < needle.Length; j++) {
                if (haystack[i + j] != needle[j]) {
                    break;
                }
            }
            if (j == needle.Length) {
                return i;
            }
        }
        return -1;
    }
    
    // TODO: add KMP
}
